core/templates/expressions/expression-syntax-tree.service.ts

```diff
@@ -16,417 +16,447 @@
-     * can be system or user defined functions and parameters, as well as
-     * system operators.
-     * @param {string} name The name to look up.
-     * @param {!Array.<!object>} envs Represents a nested name space
-     *     environment to look up the name in. The first element is looked up
-     *     first (i.e. has higher precedence).
-     * @throws {ExprUndefinedVarError} The named variable was not found in
-     *     the given environment.
-     */
-    const lookupEnvs = (name: string, envs: object[]) => {
-      for (const env of envs) {
-        if (env.hasOwnProperty(name)) {
-          return env[name];
-        }
+  public applyFunctionToParseTree(parsed: string, envs: Array<object>,
+      func: (parsed: string, envs: Array<object>) => string) {
+    return func(parsed, envs.concat(this.system));
+  }
+
+  /**
+   * Looks up a variable of the given name in the env. Here the variable
+   * can be system or user defined functions and parameters, as well as
+   * system operators.
+   * @param {string} name The name to look up.
+   * @param {!Array.<!object>} envs Represents a nested name space
+   *     environment to look up the name in. The first element is looked up
+   *     first (i.e. has higher precedence).
+   * @throws {ExprUndefinedVarError} The named variable was not found in
+   *     the given environment.
+   */
+  public lookupEnvs(name: string, envs: Array<object>): string {
+    for (const env of envs) {
+      if (env.hasOwnProperty(name)) {
+        return env[name];
       }
+    }
 
-      throw new ExprUndefinedVarError(name, envs);
-    };
+    throw new ExprUndefinedVarError(name, envs);
+  }
 
-    // Coerces the argument to a Number, and throws an error if the result
-    // is NaN.
-    var _coerceToNumber = function(originalValue) {
-      var coercedValue = (+originalValue);
-      if (!isNaN(coercedValue)) {
-        return coercedValue;
-      }
-      throw new ExprWrongArgTypeError(
-        originalValue, typeof originalValue, 'Number');
-    };
+  // Coerces the argument to a Number, and throws an error if the result
+  // is NaN.
+  private _coerceToNumber(originalValue: string): number {
+    var coercedValue = +originalValue;
+    if (!isNaN(coercedValue)) {
+      return coercedValue;
+    }
+    throw new ExprWrongArgTypeError(originalValue, typeof originalValue,
+      'Number');
+  }
 
-    // Coerces all values in the given argument array to Number, and throws
-    // an error if the result is NaN.
-    var _coerceAllArgsToNumber = function(args) {
-      for (var i = 0; i < args.length; i++) {
-        args[i] = _coerceToNumber(args[i]);
-      }
-      return args;
-    };
+  // Coerces all values in the given argument array to Number, and throws
+  // an error if the result is NaN.
+  private _coerceAllArgsToNumber(args: Array<any>): Array<number> {
+    // NOTE TO DEVELOPERS: Setting Array<number|string> and expecting
```

Comment:
Let's write this clearly. Something like:
```
Setting args to Array<number|string> throws this error (mention the error) and this._coerceToNumber(args[i]) since _coerceToNumber expect args to be just strings.
```

Reply:
Done

Reply:
@h-ARTS If you think, that it's done and won't require any further comment for the reviewer you can go ahead and resolve this particular comment.